src/lib/pubkey/classic_mceliece/cmce_poly.cpp

```diff
@@ -0,0 +1,175 @@
+
+   for(size_t i = 0; i < m_t; ++i) {
+      for(size_t j = 0; j < m_t; ++j) {
+         prod.at(i + j) += (a.coef_at(i) * b.coef_at(j));
+      }
+   }
+
+   for(size_t i = (m_t - 1) * 2; i >= m_t; --i) {
+      for(auto& [idx, coef] : m_position_map) {
+         prod.at(i - m_t + idx) += coef * prod.at(i);
+      }
+   }
+
+   prod.erase(prod.begin() + m_t, prod.end());
+
+   return Classic_McEliece_Polynomial(prod);
+}
+
+Classic_McEliece_Polynomial Classic_McEliece_Polynomial_Ring::create_element_from_bytes(
+   std::span<const uint8_t> bytes) const {
+   BOTAN_ARG_CHECK(bytes.size() == m_t * 2, "Correct input size");
+   auto coef = load_le_gf_vec(bytes);
+   return create_element_from_coef(coef);
+}
+
+Classic_McEliece_Polynomial Classic_McEliece_Polynomial_Ring::create_element_from_coef(
+   const std::vector<GF_Elem>& coeff_vec) const {
+   std::vector<Classic_McEliece_GF> coeff_vec_gf;
+   std::transform(coeff_vec.begin(), coeff_vec.end(), std::back_inserter(coeff_vec_gf), [this](auto& coeff) {
+      return Classic_McEliece_GF(coeff, m_poly_f);
+   });
+   return Classic_McEliece_Polynomial(coeff_vec_gf);
+}
+
+bool operator==(const Classic_McEliece_Polynomial_Ring::Big_F_Coefficient& lhs,
+                const Classic_McEliece_Polynomial_Ring::Big_F_Coefficient& rhs) {
+   return lhs.coeff == rhs.coeff && lhs.idx == rhs.idx;
+}
+
+std::optional<Classic_McEliece_Minimal_Polynomial> Classic_McEliece_Polynomial_Ring::compute_minimal_polynomial(
+   std::span<const uint8_t> seed) const {
+   auto polynomial = create_element_from_bytes(seed);
+   std::vector<Classic_McEliece_Polynomial> mat;
+
+   mat.push_back(create_element_from_coef(concat_as<std::vector<GF_Elem>>(
+      std::vector<GF_Elem>{GF_Elem(1)}, std::vector<GF_Elem>(degree() - 1, GF_Elem(0)))));
+
+   mat.emplace_back(polynomial);
+
+   for(size_t j = 2; j <= degree(); ++j) {
+      mat.push_back(multiply(mat.at(j - 1), polynomial));
+   }
+
+   // Gaussian
+   for(size_t j = 0; j < degree(); ++j) {
+      for(size_t k = j + 1; k < degree(); ++k) {
+         auto cond = GF_Mask::is_zero(mat.at(j).coef_at(j));
+
+         for(size_t c = j; c < degree() + 1; ++c) {
+            mat.at(c).coef_at(j) += cond.if_set_return(mat.at(c).coef_at(k));
+         }
+      }
+
+      if(mat.at(j).coef_at(j).is_zero()) {
+         // Fail if not systematic.
+         return std::nullopt;
+      }
+
+      auto inv = mat.at(j).coef_at(j).inv();
+
+      for(size_t c = j; c < degree() + 1; ++c) {
+         mat.at(c).coef_at(j) *= inv;
+      }
+
+      for(size_t k = 0; k < degree(); ++k) {
+         if(k != j) {
+            auto t = mat.at(j).coef_at(k);
+
+            for(size_t c = j; c < degree() + 1; ++c) {
+               mat.at(c).coef_at(k) += mat.at(c).coef_at(j) * t;
+            }
+         }
+      }
+   }
+
+   auto minimal_poly_coeffs = mat.at(degree()).coef();
+   // Add coefficient 1 since polynomial is monic
```

Comment:
Do we need an assertion for this?

Reply:
We don't. A monic polynomial is a polynomial of the form: $x^n+c_{n-1}x^{n-1} + ... + c_1x^1 + c_0$. The above algorithm computes $c_0,...,c_{n-1}$. Therefore, to compose the final polynomial, we add $c_n = 1$ to represent the additional monomial $x^n$. There is nothing to be checked here.

Reply:
d'accord.